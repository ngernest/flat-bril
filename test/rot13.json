{
  "functions": [
    {
      "args": [
        {
          "name": "input",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "input"
          ],
          "dest": "aed",
          "funcs": [
            "rot"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "aed"
          ],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "thirteen",
          "op": "const",
          "type": "int",
          "value": 13
        },
        {
          "dest": "twenty_six",
          "op": "const",
          "type": "int",
          "value": 26
        },
        {
          "args": [
            "n",
            "thirteen"
          ],
          "dest": "shifted",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "shifted",
            "twenty_six"
          ],
          "dest": "over",
          "op": "ge",
          "type": "bool"
        },
        {
          "args": [
            "over"
          ],
          "labels": [
            "if",
            "else"
          ],
          "op": "br"
        },
        {
          "label": "if"
        },
        {
          "args": [
            "shifted",
            "twenty_six"
          ],
          "dest": "shifted",
          "op": "sub",
          "type": "int"
        },
        {
          "label": "else"
        },
        {
          "args": [
            "shifted"
          ],
          "op": "ret"
        }
      ],
      "name": "rot",
      "type": "int"
    }
  ]
}
