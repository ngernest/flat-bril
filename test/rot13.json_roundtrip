{
  "functions": [
    {
      "args": [
        {
          "name": "input",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "input"
          ],
          "dest": "aed",
          "funcs": [
            "rot"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "aed"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "thirteen",
          "op": "const",
          "type": "int",
          "value": 13
        },
        {
          "dest": "twenty_six",
          "op": "const",
          "type": "int",
          "value": 26
        },
        {
          "args": [
            "n",
            "thirteen"
          ],
          "dest": "shifted",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "shifted",
            "twenty_six"
          ],
          "dest": "over",
          "funcs": [],
          "labels": [],
          "op": "ge",
          "type": "bool"
        },
        {
          "args": [
            "over"
          ],
          "funcs": [],
          "labels": [
            "if",
            "else"
          ],
          "op": "br"
        },
        {
          "label": "if"
        },
        {
          "args": [
            "shifted",
            "twenty_six"
          ],
          "dest": "shifted",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "label": "else"
        },
        {
          "args": [
            "shifted"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "rot",
      "type": "int"
    }
  ]
}
