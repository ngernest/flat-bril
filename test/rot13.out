@main
	op: call 		dest: aed		type: int		args: ["input"]		funcs: ["rot"]
	op: print		args: ["aed"]	
@rot
	op: const		dest: thirteen		type: int		value: 13	
	op: const		dest: twenty_six		type: int		value: 26	
	op: add  		dest: shifted		type: int		args: ["n", "thirteen"]	
	op: ge   		dest: over		type: bool		args: ["shifted", "twenty_six"]	
	op: br   		args: ["over"]		labels: ["if", "else"]
.if
	op: sub  		dest: shifted		type: int		args: ["shifted", "twenty_six"]	
.else
	op: ret  		args: ["shifted"]	
